apps/flame-defi/app/features/evm-wallet/services/token-approval-service/token-approval-service.ts

```diff
@@ -0,0 +1,83 @@
+import type { Config } from "@wagmi/core";
+import { type Address, erc20Abi } from "viem";
+import { GenericContractService } from "../generic-contract-service";
+import { EvmCurrency } from "@repo/flame-types";
+
+export class TokenApprovalService extends GenericContractService {
+  constructor(wagmiConfig: Config, contractAddress: Address) {
+    super(wagmiConfig, contractAddress, erc20Abi);
+  }
+
+  async approveToken(
+    chainId: number,
+    userAddress: `0x${string}`,
+    contractAddress: Address,
+    token: EvmCurrency | null | undefined,
+    tokenAllowances: { symbol: string; allowance: bigint }[],
+    tokenDefaultApprovalAmount: string,
+  ) {
+    const amountAsBigInt = BigInt(tokenDefaultApprovalAmount);
+    // NOTE: Reset this to 0 whenever we want to reset the approval
+    // const amountAsBigInt = BigInt('0');
+
+    try {
+      const txHash = await this.writeContractMethod(
+        chainId,
+        "approve",
+        [contractAddress, amountAsBigInt],
+        { account: userAddress },
+      );
+
+      const newTokenAllowance = tokenAllowances.map((data) => {
+        if (data.symbol === token?.coinDenom) {
+          return {
+            symbol: token?.coinDenom,
+            allowance: BigInt(tokenDefaultApprovalAmount),
+          };
+        }
+        return data;
+      });
+
+      return { allowances: newTokenAllowance, txHash };
+    } catch (error) {
+      console.warn("Failed to approve token:", error);
+      return null;
```

Comment:
I think we should do the try/catch in the parent only. I do use `try` in the generic service, but I rethrow the error. In general I think errors should be handled closer to the code that is doing something with that error, e.g. displaying an error modal.